build/packer: Change regex expressions used by Filebeat to process log events

spiffyy99 opened the pull request:
build/packer: Change Filebeat regexes to recognize releases of Helm
charts and libgeos files in log messages.

Release Note: None

knz committed:
heapprofiler: properly preserve heap dumps

Prior to this patch, the heap profiler GC was very aggressive and also
incorrect.

Agressive, because it was deleting every file in excess of the
configured max_profiles (5 entries), regardless of utility.

This is annoying because if a node crashes due to an OOM error
and then is restarted, it may take a while for an operator to
notice. By the time they notice, the GC would already have removed the
"interesting" files from the time of the crash.

It was also incorrect, because it was improperly considering
"unexpected" files (e.g. files from a previous version, from other
programs) in the GC process, whereby even fewer legit files would have
a chance to remain.

This patch addresses this by enhancing the GC process as follows:

- it now understands "ramp ups": a sequence of heap dumps where
  the heap was increasing in size.

- it only applies the "max profiles" limit to the files that are part
  of the *last ramp up*: the most recent sequence of heap dumps where
  the heap size is monotonically increasing.

  Files that are not part of this group are not considered for
  deletion from the "max profiles" limit (but see below).

  This ensures that after an OOM crash, the dumps pertaning to that
  crash are not immediately deleted by the new instance of the
  process.

- it introduces a size-based retention policy: all the most recent
  files that fit within a configurable maximum are retained, regardless
  of age. The default maximum is 64MB, which corresponds to ~600
  profiles according to observation.

  The algorithm is careful to ensure that the last "max profiles" from
  the last ramp-up are preserved regardless of age, even if they
  exceed the configured max duration.

Release note (cli change): CockroachDB now better attempts to retain
heap profile dumps after a crash due to an out-of-memory error.

Release note (backward-incompatible change): The file names for heap
profile dumps are now `memprof.<date-and-time>.<heapsize>`, where
previously they were named `memprof.<heapsize>.<date-and-time>`.

knz committed:
server: properly name a parameter

(a period is not a frequency)

Release note: None

knz committed:
server: simplify the metric sampling initialization

The code was already able to inhibit sampling if there was no output
directory; no need to look at the list of configured stores to confirm
what is already known.

Release note: None

knz committed:
server/{goroutinedumper,heapprofiler}: don't delete unknown files

Previously, the code was deleting files that it did not generate.
This was dangerous as the user may have legitimate reasons to
redirect dumps to a different directory with perhaps other data in it.

Release note: None

knz committed:
server: share GC code between heap profiler and goroutine dumper

since the GC logic should remain the same (and is non-trivial), this
commit removes the code duplication by creating a common component `dumpstore`.

Release note: None

knz committed:
server,start: share code for Go heap and jemalloc profile collection

Prior to this patch, there were two separate loops to collect heap
profiles:

- one started by the periodic metric collection by
  `server.Server` (via `startSampleEnvironment()`),
  running every 10s, with a heuristic to minimize collections
  when unneeded, and a minimum sample frequency of 1h (i.e ignore the
  heuristic at least once an hour).

- another in `cli/start.go`, triggered upon setting
  `COCKROACH_MEMPROF_INTERVAL` (disabled by default).
  This loop was additionally collecting jemalloc profiles.

Both were collecting in the same directory with similarly named files
but not exactly. Their garbage collectors were also getting in the way
of each other.

This patch improves upon this by bringing the 2nd loop into a special
case of the first: if `COCKROACH_MEMPROF_INTERVAL` is set, it
overrides the minimum sample frequency of 1h.

The new behavior is thus to collect at least every hour, now
customizable via `COCKROACH_MEMPROF_INTERVAL`; with automatic
additional collection when there are memory usage increases (the
previous heuristic).

If `COCKROACH_MEMPROF_INTERVAL` is set to zero or negative, or if
there is no logging directory, heap collection to disk is entirely
disabled (as previously).

The common loop also collects jemalloc and Go heap profiles
side-by-side. Their threshold ("memory usage") values are:

- for Go heap allocs: memstats `GoAllocBytes`
- for Non-go allocs: memstats `CGoTotalBytes`

Release note (backward-compatible change): The reserved,
non-documented cluster settings `server.heap_profile.xxx` have been
renamed to `server.mem_profile.xxx`. They now control collection of
multiple sorts of memory profiles besides just Go heap allocations.

knz committed:
cli: move the CPU profiling code to its own file

Release note: None

knz committed:
cli/cpuprofile: reuse the same GC policy as the other profilers

The profile dump GC code is now shared between goroutine dumps, heap
profiles and jemalloc profiles. This patch does one more round of
sharing and subsumes the CPU profile collector under the same
mechanism.

Release note: None

knz committed:
server,cli: keep the GC silent about unknown files

The previous iteration on the GC code was complaining when it
encountered unknown files; this was useful for debugging, but is a bit
noisy for e.g. the CPU profiling which is going to the same
directory as log files by default.

Release note: None

Commit:
bulkio: Correctly handle exhausting retries when reading from HTTP.

Return an error if we exhaust the retry budget when reading from HTTP.

Release Notes: None

rohany committed:
sql: implement `ALTER SCHEMA RENAME TO`

Fixes #50880.

This commit implements the `ALTER SCHEMA RENAME TO` command. It reuses
the existing schema change infrastructure to handle a more general case
of an arbitrary descriptor that just needs names and existing leases
drained.

Release note (sql change): Implement the `ALTER SCHEMA RENAME TO`
command.

yuzefovich committed:
colexec: add default comparison operators

This commit introduces default comparison projection and selection
operators that handle all `tree.ComparisonExpr`s for which we don't
have optimized implementations. The main operators are very similar
to `defaultBuiltinFuncOperator`, but we also introduce optimized
adapter implementations from `tree.ComparisonExpr` to a vectorized
friendly model. Quick benchmarks show about 3.5x improvement in
speed of `IS DISTINCT FROM` projection operator against the wrapped
post-processing spec.

Release note (sql change): Vectorized execution engine now fully
supports comparison operators (things like `ILIKE`, `IS NOT DISTINCT
FROM`, `SIMILAR TO`, and several others).

irfansharif committed:
roachtest: deflake decommissioning roachtest

Fixes #52304. Also rename this test so it gets grouped with the other
decommissioning-related roachtests in the generated artifacts.

Release note: None

yuzefovich committed:
colexec: add support for Tuple expressions and clean up tests

This commit adds support for `tree.Tuple` expressions which are
supported either by pre-evaluation during planning (and handling the
result as other constant values) if the tuple is constant or by planning
projection operators for each expression in the tuple and then using
newly-introduced `tupleProjOp` if the tuple expression is not constant.

The support required refactoring of IsNull* operators to template out special
"is tuple null" variants (because tuple have very peculiar null-handling
semantics). It also required that we fallback to the default comparison
operators on things that we have optimized support for (for example,
EQ), also because of the null-handling semantics.

It also improves the tests to handle datum-backed types better. Note
that a single unit test in `aggregators_test.go` has been switched to use
`types.TimeTZ` instead of `types.Jsonb` because of the differences in
order of `string`s and `json.JSON.String()` (the former is the
"expected" value whereas the latter is the actual; it was simpler to
switch the test to use a different datum-backed type).

Release note: None

knz committed:
server/heapprofiler: collect general memory stats upon RSS increases

The memory profiler intends to collect profiles when "memory usage
is increasing". Prior to this patch, it was only looking at
Go *heap* usage and non-Go allocs in this heuristic. This was
unfortunate, because in practice we know that processes fail with
OOM errors also for other allocations than just heap
usage (e.g. Go stacks, goroutines, etc.).

To make the heap profiler more useful, this patch adds a new
general memory stats dumper, which uses current process RSS
as a threshold value to generate output files.

Output files are named `memstats.xxx` and are emitted alongside
`memprof.xxx` and `jeprof.xxx` files in the same output directory.
They contain:

- a JSON copy of the `status.GoMemStats` struct
- underneath, a JSON copy of the (new) `status.CgoMemStats`

Additionally, the RSS total size is populated in the file name.

Release note (cli change): CockroachDB now better attempts to retain
memory statistics corresponding to increases in total memory usage,
not just heap allocations.

rohany committed:
backupccl: enable backing up and restoring user defined schemas

Fixes #50875.

This commit adds support for backup/restore to operate when user defined
schemas have been created. We don't add a new `BACKUP SCHEMA` command,
as this would most likely create backups that have dangling references
across schemas.

Release note (enterprise change): Enable backup and restore when user
defined schemas are in use.

jramos-crl committed:
authors: add jramos-crl to authors

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52646)
<!-- Reviewable:end -->

jlinder reviewed (commented):
Looking good. Let's get the image built.

DrewKimball committed:
opt: streamline stats test rewrite process

This PR pulls the `save-tables` and `stats` opttester commands into
a single command: `stats-quality`. Stats quality tests can now be
rewritten without having to manually remove `stats` tests beforehand,
or having to manually add them back afterword.

By default, stats are outputted for every expression in the plan tree.
I have added a flag: `ignore-tables` that allows a set of stats tables
to be specified for which stats should not be outputted. Note that
care should be taken when rewritting stats for a test that uses the
`ignore-tables` flag, since the expressions that are being 'ignored'
may change if the plan changes.

Release note: None

Commit:
flowinfra: replace ConcurrentExecution with ConcurrentTxnUse

Previously, the DistSQLPlanner would reject any local flows with any
concurrency due to possible concurrent use of a RootTxn. This policy was overly
prohibitive as it would reject stats flows that contain processors that cannot
be fused even though there would be no concurrent txn usage.

This commit makes this policy a bit more lenient, introducing a way for
processors to indicate that they and their input will not be using a txn. This
is exposed through the new ConcurrentTxnUse method in the flow interface, which
returns whether more than one processor will be using a txn.

This commit also removes ConcurrentExecution since all usages have been
replaced with ConcurrentTxnUse.

Release note: None (internal refactor to fix an error message received only in
a multitenant environment)

dt committed:
jobs: add job type to state machine logging

Release note: none.

knz committed:
cli/demo: activate Sentry crash reporting for `cockroach demo`

Prior to this patch, the TestServer instance was using its own
`cluster.Settings` instance separate from the global one defined in
`cli.serverCfg`, which is the one used by the `log` package to decide
whether or not to send sentry reports.

Since the `Settings` objects did not match, sentry reporting never got
enabled for `cockroach demo`.

This patch fixes it.

This can be checked by running `cockroach demo --logtostderr=INFO`,
with `COCKROACH_CRASH_REPORTS` set to a suitable URL. The reporting
will be visible as the following log line (example):

```
E200812 16:27:10.520228 1764 util/log/crash_reporting.go:319  [n1,client=127.0.0.1:33741,hostssl,user=root] Queued as error 694cfa684ea14a528be7de9e79aa5947
```

Release note (cli change): Crashes in `cockroach demo` sessions are
now reported to telemetry, if telemetry is enabled.

Commit:
server: don't attempt to upgrade the version until the SQL server starts

This attempt to upgrade the server prior to the sql server starting means that
we might mark a version as active before the associated sqlmigration has been
performed, which can cause big problems.

Release note: None

knz committed:
util/log: indicate when sentry reports come from `log.Fatal`

This commit adds a prefix `log.Fatal` to the error message string,
and also changes the `report_type` sentry tag to be `log_fatal`.

Release note: None

yuzefovich committed:
sql/col*: speed up disk queue tests

Disk queue tests make sure that batches that are read from disk are
equivalent to the batches that have been written down. The equivalence
is checked via a string-matching, and stringifying can take non-trivial
amount of time. Previously, the tests would create up to 1024 batches,
and if random operator chose `types.Geography` as one of the random
types, stringifying would take long time timing out the test. This
commit reduces that number to 16.

Release note: None

knz committed:
pkg/server: minor code simplification

Release note: None

RaduBerinde committed:
sql: fix TestTelemetry flake

This test started showing `sql.plan.cte` for an unrelated query. I am
guessing we added some background query that uses CTEs. Improving the
test to ignore this counter outside of the specific testcase that
checks for it.

Fixes #52679.

Release note: None

Commit:
Add get() to ttl_seconds to prevent exception

In a reported case from a customer, `zone_config.gc` is undefined and
accessing `ttl_seconds` on a null threw an error that stopped page
rendering.

I'm adding a lodash/get call here with `null` as a default return to
prevent a TypeError. This is generally a good idea whenever you are
accessing anything that could be undefined in javascript.

Release note: None

rohany committed:
importccl: disable import with user defined schemas

Release note (enterprise change): Cannot use `IMPORT` to create tables
in user defined schemas. Users should instead create the table with
`CREATE TABLE` and then use `IMPORT INTO`.

irfansharif committed:
roachtest: skip acceptance/version-upgrade

This has been failing CI, and is being tracked in #52627. This is a
pretty critical test to have so we'll be addressing it shortly.

Release note: None

otan committed:
geomfn: add handling of 0-2 in RelateMatch

Release note: None

dt committed:
kvserver: add deadline to ClearRange

In many cases we would like to quickly and cheaply empty a key-range
and then make it available for SQL traffic. ClearRange is good at
quickly and cheaply emptying a key-range, but currently fears that
a copy of a ClearRange request could get replayed later make us fear
using it on a span which will later be sent writes -- if that replay
were to happen after writes resume, we could accidentally clear those
writes.

This change adds a Deadline parameter to ClearRange which the caller
can set to a timestamp to instruct the server to refuse to execute the
ClearRange request at or after that time. Thus if a caller sends the
ClearRanges with some deadline X, it can then wait until X (plus the
max clock offset) and then resume writing to the span it cleared. If
for any reason one of its ClearRange requests got stuck and a copy
pops out of a buffer after the deadline, it will have no effect.

Release note: none.

otan committed:
Makefile: bump lint timeout to 30m

We have too much code that the linter needs more than 20 minutes these
days (the ones that don't take ~19). Better make that 30.

Release note: None

spiffyy99 committed:
Change regex expressions used by Filebeat to process log events

build/packer: Change Filebeat regexes to recognize releases of Helm
charts and libgeos files in log messages.

Release Note: None

spiffyy99 committed:
Change regex expressions used by Filebeat to process log events

build/packer: Change Filebeat regexes to recognize releases of Helm
charts and libgeos files in log messages.

Release Note: None

Commit:
Merge #52690

52690: jobs: add job type to state machine logging r=dt a=dt

Release note: none.

Co-authored-by: David Taylor <tinystatemachine@gmail.com>

Commit:
Merge #52713 #52735 #52737

52713: sql: fix TestTelemetry flake r=yuzefovich a=RaduBerinde

This test started showing `sql.plan.cte` for an unrelated query. I am
guessing we added some background query that uses CTEs. Improving the
test to ignore this counter outside of the specific testcase that
checks for it.

Fixes #52679.

Release note: None

52735: geomfn: add handling of 0-2 in RelateMatch r=sumeerbhola a=otan

Release note: None

52737: Makefile: bump lint timeout to 30m r=irfansharif a=otan

We have too much code that the linter needs more than 20 minutes these
days (the ones that don't take ~19). Better make that 30.

Resolves #52264.

Release note: None

Co-authored-by: Radu Berinde <radu@cockroachlabs.com>
Co-authored-by: Oliver Tan <otan@cockroachlabs.com>

RaduBerinde committed:
opt: store columns in OpaqueMetadata

Store the result columns of opaque operators in their metadata instead
of returning it separately.

Release note: None

RaduBerinde committed:
opt: separate CreateTable/CreateTableAs exec operators

Separate the two variants in the exec factory so that we don't have to
pass around a `nil` node.

Release note: None

RaduBerinde committed:
sql: new EXPLAIN infrastructure

This change reimplements the default `(PLAN)` variant of EXPLAIN.

Instead of visiting a plan node tree to get EXPLAIN information, we
record the exact `exec.Factory` calls that were produced. We do this
using an `exec.ExplainFactory` which is an extension of
`exec.Factory`.

This factory wraps a regular `exec.Factory` and forwards all factory
calls to it (while also recording all arguments). The result is that
for each `exec.Node` there is a corresponding `*explain.Node`. The
factory also provides a way for the execbuilder to annotate nodes with
more information, which will be initially used for row count and cost
(not exposed in this PR).

The implementation relies on optgen for the factory methods and
per-operation identifiers and structs.

I tried to avoid major changes to the EXPLAIN output; I left a lot of
TODOs for improvements that I plan to make in follow-up changes. Some
egregious issues were fixed in separate PRs. Still, there are various
differences; the notable ones are:
 - for simple projections, we now show a `project` node without any
   rendering expressions.
 - in some cases we see a new `project` that just rearranges columns.
   I will address this in a follow-up change.
 - window function information is no longer duplicated.
 - expressions for inverted filters and joins now show the column
   names instead of unnamed references like `@2`.
 - we no longer see a `count` node on top of mutations.

Release note (sql change): various improvements to EXPLAIN.

RaduBerinde committed:
sql: EXPLAIN test updates

Separate commit for explain test changes.

Release note: None

yuzefovich committed:
distsql: fix TestAggregatorAgainstProcessor

In `TestAggregatorAgainstProcessor` we generate input types randomly
relying on `execinfrapb.GetAggregateInfo` function to perform the
"type-checking". That function, however, can have "false negative" for
several aggregate functions when a tuple is chosen as an input type, so
we have special handling for those cases.

Release note: None

yuzefovich committed:
rowexec: fix closing of aggregator bucket in case of an error

Previously, the bucket of aggregate functions wouldn't get properly
closed if an aggregate function returned an error on `Result` call. Now
this is fixed.

Release note: None

Commit:
Merge #52732

52732: roachtest: skip acceptance/version-upgrade r=irfansharif a=irfansharif

This has been failing CI, and is being tracked in #52627. This is a
pretty critical test to have so we'll be addressing it shortly.

Release note: None

Co-authored-by: irfan sharif <irfanmahmoudsharif@gmail.com>

Commit:
Merge #52520

52520: roachtest: deflake decommissioning roachtest r=irfansharif a=irfansharif

Fixes #52304. Also rename this test so it gets grouped with the other
decommissioning-related roachtests in the generated artifacts.

Release note: None

Co-authored-by: irfan sharif <irfanmahmoudsharif@gmail.com>

adityamaru committed:
parser: add DETACHED option to run RESTORE in explicit txn

Add DETACHED option to the list of allowed RESTORE options.  This option
requests the restore to run in detached mode: that is, the restore
statement should not block, and instead simply return the job id.

Release note: None

Commit:
Merge #52697

52697: server: don't attempt to upgrade the version until the SQL server starts r=irfansharif a=ajwerner

This attempt to upgrade the server prior to the sql server starting means that
we might mark a version as active before the associated sqlmigration has been
performed, which can cause big problems.

Release note: None

Co-authored-by: Andrew Werner <ajwerner@cockroachlabs.com>

Commit:
Merge #52621

52621: flowinfra: replace ConcurrentExecution with ConcurrentTxnUse r=yuzefovich a=asubiotto

Previously, the DistSQLPlanner would reject any local flows with any
concurrency due to possible concurrent use of a RootTxn. This policy was overly
prohibitive as it would reject stats flows that contain processors that cannot
be fused even though there would be no concurrent txn usage.

This commit makes this policy a bit more lenient, introducing a way for
processors to indicate that they and their input will not be using a txn. This
is exposed through the new ConcurrentTxnUse method in the flow interface, which
returns whether more than one processor will be using a txn.

This commit also removes ConcurrentExecution since all usages have been
replaced with ConcurrentTxnUse.

Release note: None (internal refactor to fix an error message received only in
a multitenant environment)

Fixes #50049 

Co-authored-by: Alfonso Subiotto Marques <alfonso@cockroachlabs.com>

Commit:
logictest: run SQLLite logic tests with 3node-tenant

We used to run this configuration as part of the default configurations, but
since the change to CCL, 3node-tenant is no longer a default configuration.

This commit adds TestTenantSQLLiteLogic, which runs the SQLLite logic tests
with a CCL binary and the 3node-tenant configuration. This test is also added
to build/teamcity-sqllogictest.sh to run nightly.

Release note: None

Commit:
logictest: swap directives with default config when possible

A while back, we added the ability to run logic tests without a directive,
which would run all default configs.

Some tests still specified explicit configs even though they could run without
a problem with all default configs. Even though it takes more time, running
with all default configs is preferable, since developers rely on the widespread
usage of the default config for good test coverage of new features.

Additionally, all files affected can also be run in a multitenant configuration
using 3node-tenant. So that TestTenantLogic may run these files, this commit
adds a blocklist directive with an issue number where appropriate for any tests
that fail in 3node-tenant that previously had explicit directives.

Release note: None (testing change)

Commit:
Merge #52677

52677: logictest: run SQLLite logic tests with 3node-tenant r=yuzefovich,nvanbenschoten a=asubiotto

We used to run this configuration as part of the default configurations, but
since the change to CCL, 3node-tenant is no longer a default configuration.

This commit adds TestTenantSQLLiteLogic, which runs the SQLLite logic tests
with a CCL binary and the 3node-tenant configuration. This test is also added
to build/teamcity-sqllogictest.sh to run nightly.

Release note: None

Co-authored-by: Alfonso Subiotto Marques <alfonso@cockroachlabs.com>

knz committed:
tests: skip distsql/TestAggregatorAgainstProcessor

See #52742.

Release note: None

knz committed:
jobs: skip TestRegistryLifecycle/rollback

Release note: None

Commit:
Merge #50446

50446: heapprofiler: properly preserve heap dumps, refactor some code r=tbg a=knz

Fixes #50237 

See the individual commits for details.

Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>

Commit:
Replacing get() with Typescript optional chain operator

As of TypeScript 3.7, the optional chaining operator provides roughly
the same purpose as lodash/get (with better syntax in my opinion). I am
removing get() and instead depending on an optional chain.

There was also concern about the correctness of displaying
`gc.ttlseconds` as a null value, which is valid. The error I am seeking
to prevent here is that `gc` is null or undefined, and so if there is no
`gc` value then I am instead choosing not to display a value at all.

Release note (bug fix): don't display a value for gc.ttlseconds if not
set

Commit:
Merge #52281 #52696 #52749

52281: bulkio: Correctly handle exhausting retries when reading from HTTP. r=rohany a=miretskiy

Fixes #52279

Return an error if we exhaust the retry budget when reading from HTTP.

Release Notes: None

52696: cli/demo: activate Sentry crash reporting for `cockroach demo` r=irfansharif a=knz

Prior to this patch, the TestServer instance was using its own
`cluster.Settings` instance separate from the global one defined in
`cli.serverCfg`, which is the one used by the `log` package to decide
whether or not to send sentry reports.

Since the `Settings` objects did not match, sentry reporting never got
enabled for `cockroach demo`.

This patch fixes it.

This can be checked by running `cockroach demo --logtostderr=INFO`,
with `COCKROACH_CRASH_REPORTS` set to a suitable URL. The reporting
will be visible as the following log line (example):

```
E200812 16:27:10.520228 1764 util/log/crash_reporting.go:319  [n1,client=127.0.0.1:33741,hostssl,user=root] Queued as error 694cfa684ea14a528be7de9e79aa5947
```

Release note (cli change): Crashes in `cockroach demo` sessions are
now reported to telemetry, if telemetry is enabled.

52749: rowexec,distsql: fix a couple of things around aggregators r=yuzefovich a=yuzefovich

**rowexec: fix closing of aggregator bucket in case of an error**

Previously, the bucket of aggregate functions wouldn't get properly
closed if an aggregate function returned an error on `Result` call. Now
this is fixed.

Release note: None

**distsql: fix TestAggregatorAgainstProcessor**

In `TestAggregatorAgainstProcessor` we generate input types randomly
relying on `execinfrapb.GetAggregateInfo` function to perform the
"type-checking". That function, however, can have "false negative" for
several aggregate functions when a tuple is chosen as an input type, so
we have special handling for those cases.

Fixes: #52742.

Release note: None

Co-authored-by: Yevgeniy Miretskiy <yevgeniy@cockroachlabs.com>
Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>
Co-authored-by: Yahor Yuzefovich <yahor@cockroachlabs.com>

Commit:
Merge #52730

52730: sql: new EXPLAIN infrastructure r=RaduBerinde a=RaduBerinde

#### opt: store columns in OpaqueMetadata

Store the result columns of opaque operators in their metadata instead
of returning it separately.

Release note: None

#### opt: separate CreateTable/CreateTableAs exec operators

Separate the two variants in the exec factory so that we don't have to
pass around a `nil` node.

Release note: None

#### sql: new EXPLAIN infrastructure

This change reimplements the default `(PLAN)` variant of EXPLAIN.

Instead of visiting a plan node tree to get EXPLAIN information, we
record the exact `exec.Factory` calls that were produced. We do this
using an `exec.ExplainFactory` which is an extension of
`exec.Factory`.

This factory wraps a regular `exec.Factory` and forwards all factory
calls to it (while also recording all arguments). The result is that
for each `exec.Node` there is a corresponding `*explain.Node`. The
factory also provides a way for the execbuilder to annotate nodes with
more information, which will be initially used for row count and cost
(not exposed in this PR).

The implementation relies on optgen for the factory methods and
per-operation identifiers and structs.

I tried to avoid major changes to the EXPLAIN output; I left a lot of
TODOs for improvements that I plan to make in follow-up changes. Some
egregious issues were fixed in separate PRs. Still, there are various
differences; the notable ones are:
 - for simple projections, we now show a `project` node without any
   rendering expressions.
 - in some cases we see a new `project` that just rearranges columns.
   I will address this in a follow-up change.
 - window function information is no longer duplicated.
 - expressions for inverted filters and joins now show the column
   names instead of unnamed references like `@2`.
 - we no longer see a `count` node on top of mutations.

Release note (sql change): various improvements to EXPLAIN.

#### sql: EXPLAIN test updates

Separate commit for explain test changes.

Release note: None

Co-authored-by: Radu Berinde <radu@cockroachlabs.com>

adityamaru committed:
backupccl: support running RESTORE in explicit txn

Allow execution of RESTORE statement under explicit transaction,
provided the "DETACHED" option is specified:

RESTORE FROM <location_uri> ... WITH DETACHED When restore runs in
detached mode, we do not wait for the job completion. Instead, we return
the job id to the caller, and the job executes asynchronously.

Release note (enterprise change): RESTORE can now run in "detached"
mode.  That is, we do not wait for the RESTORE job to finish. Instead,
the job ID is returned immediately, and the job itself runs in
background.

Commit:
Merge #52729

52729: importccl: disable import with user defined schemas r=rohany a=rohany

Release note (enterprise change): Cannot use `IMPORT` to create tables
in user defined schemas. Users should instead create the table with
`CREATE TABLE` and then use `IMPORT INTO`.

Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>

yuzefovich committed:
distsql: unskip TestAggregatorAgainstProcessor

The test has been fixed.

Release note: None

Commit:
Merge #52641

52641: authors: add jramos-crl to authors r=otan a=jramos-crl

Release note: None

Co-authored-by: Jose <jose@cockroachlabs.com>

Commit:
bulkio: Require admin role when creating backup schedules.

Require admin role when creating backup schedules.

Release Notes: None

Commit:
Merge #52772

52772: distsql: unskip TestAggregatorAgainstProcessor r=yuzefovich a=yuzefovich

The test has been fixed.

Release note: None

Co-authored-by: Yahor Yuzefovich <yahor@cockroachlabs.com>

Commit:
Merge #52680

52680: logictest: swap directives with default config when possible r=yuzefovich a=asubiotto

A while back, we added the ability to run logic tests without a directive,
which would run all default configs.

Some tests still specified explicit configs even though they could run without
a problem with all default configs. Even though it takes more time, running
with all default configs is preferable, since developers rely on the widespread
usage of the default config for good test coverage of new features.

Additionally, all files affected can also be run in a multitenant configuration
using 3node-tenant. So that TestTenantLogic may run these files, this commit
adds a blocklist directive with an issue number where appropriate for any tests
that fail in 3node-tenant that previously had explicit directives.

Release note: None (testing change)

Addresses #52410 

Co-authored-by: Alfonso Subiotto Marques <alfonso@cockroachlabs.com>

Commit:
Merge #52778

52778: bulkio: Require admin role when creating backup schedules. r=miretskiy a=miretskiy

Require admin role when creating backup schedules.

Release Notes: None

Co-authored-by: Yevgeniy Miretskiy <yevgeniy@cockroachlabs.com>

Commit:
Merge #52313 #52349

52313: colexec: add default comparison operator r=yuzefovich a=yuzefovich

Depends on #52315.

**colexec: add default comparison operators**

This commit introduces default comparison projection and selection
operators that handle all `tree.ComparisonExpr`s for which we don't
have optimized implementations. The main operators are very similar
to `defaultBuiltinFuncOperator`, but we also introduce optimized
adapter implementations from `tree.ComparisonExpr` to a vectorized
friendly model. Quick benchmarks show about 3.5x improvement in
speed of `IS DISTINCT FROM` projection operator against the wrapped
post-processing spec.

Addresses: #49781.

Release note (sql change): Vectorized execution engine now fully
supports comparison operators (things like `ILIKE`, `IS NOT DISTINCT FROM`,
`SIMILAR TO`, and several others).

**colexec: add support for Tuple expressions and clean up tests**

This commit adds support for `tree.Tuple` expressions which are
supported either by pre-evaluation during planning (and handling the
result as other constant values) if the tuple is constant or by planning
projection operators for each expression in the tuple and then using
newly-introduced `tupleProjOp` if the tuple expression is not constant.

The support required refactoring of IsNull* operators to template out special
"is tuple null" variants (because tuple have very peculiar null-handling
semantics). It also required that we fallback to the default comparison
operators on things that we have optimized support for (for example,
EQ), also because of the null-handling semantics.

It also improves the tests to handle datum-backed types better. Note
that a single unit test in `aggregators_test.go` has been switched to use
`types.TimeTZ` instead of `types.Jsonb` because of the differences in
order of `string`s and `json.JSON.String()` (the former is the
"expected" value whereas the latter is the actual; it was simpler to
switch the test to use a different datum-backed type).

Release note: None

52349: sql: implement `ALTER SCHEMA RENAME TO` r=rohany a=rohany

Fixes #50880.

This commit implements the `ALTER SCHEMA RENAME TO` command. It reuses
the existing schema change infrastructure to handle a more general case
of an arbitrary descriptor that just needs names and existing leases
drained.

Release note (sql change): Implement the `ALTER SCHEMA RENAME TO`
command.

Co-authored-by: Yahor Yuzefovich <yahor@cockroachlabs.com>
Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>

Commit:
bulkio: Fix racy test

Fix a race in the test, where the previously created
backup schedules would still continue executing, creating more backup
directories, which would cause cleanup function to fail.

Release Notes: None

Commit:
Merge #52708 #52779

52708: sql/col*: speed up disk queue tests r=yuzefovich a=yuzefovich

Disk queue tests make sure that batches that are read from disk are
equivalent to the batches that have been written down. The equivalence
is checked via a string-matching, and stringifying can take non-trivial
amount of time. Previously, the tests would create up to 1024 batches,
and if random operator chose `types.Geography` as one of the random
types, stringifying would take long time timing out the test. This
commit reduces that number to 16.

Fixes: #52398.
Fixes: #52532.

Release note: None

52779: parser, backupccl: support running RESTORE in explicit txn r=miretskiy a=adityamaru

Fixes: #50735

Co-authored-by: Yahor Yuzefovich <yahor@cockroachlabs.com>
Co-authored-by: Aditya Maru <adityamaru@gmail.com>

Commit:
Merge #52699

52699: util/log: indicate when sentry reports come from `log.Fatal` r=irfansharif a=knz

First commit from #52696.

Requested by @andreimatei  in https://github.com/cockroachdb/cockroach/issues/52600#issuecomment-672944490

Before:

![image](https://user-images.githubusercontent.com/642886/90043609-9fc2bc00-dccc-11ea-95a9-008939771225.png)
![image](https://user-images.githubusercontent.com/642886/90043633-a6513380-dccc-11ea-8530-c64ad250f348.png)
![image](https://user-images.githubusercontent.com/642886/90043649-ab15e780-dccc-11ea-9778-48ab885326ff.png)

After:

![image](https://user-images.githubusercontent.com/642886/90043668-b23cf580-dccc-11ea-909b-ec59e646039d.png)
![image](https://user-images.githubusercontent.com/642886/90043674-b701a980-dccc-11ea-8718-bdb3b3b50a11.png)
![image](https://user-images.githubusercontent.com/642886/90043693-bc5ef400-dccc-11ea-9ab1-76e4e95b300c.png)

For example: https://sentry.io/organizations/cockroach-labs/issues/1837472473/

Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>

Commit:
Merge #52721

52721: Check value of gc.ttlseconds before rendering to prevent TypeError r=nathanstilwell a=nathanstilwell

In a reported case from a customer, `zone_config.gc` is undefined and
accessing `ttl_seconds` on a null threw an error that stopped page
rendering.

```
bundle.js:49 TypeError: Cannot read property 'ttl_seconds' of null
    at t.e.renderZone (bundle.js:83)
    at t.value (bundle.js:83)
    at Xi (bundle.js:49)
    at $i (bundle.js:49)
    at Ts (bundle.js:49)
    at Cc (bundle.js:49)
    at Pc (bundle.js:49)
    at wc (bundle.js:49)
    at bundle.js:49
    at t.unstable_runWithPriority (bundle.js:57)
```

I'm adding a lodash/get call here with `null` as a default return to
prevent a TypeError. This is generally a good idea whenever you are
accessing anything that could be undefined in javascript.



Co-authored-by: Nathan Stilwell <nathanstilwell@cockroachlabs.com>

Commit:
Merge #52666

52666: opt: streamline stats test rewrite process r=DrewKimball a=DrewKimball

This PR pulls the `save-tables` and `stats` opttester commands into
a single command: `stats-quality`. Stats quality tests can now be
rewritten without having to manually remove `stats` tests beforehand,
or having to manually add them back afterword.

By default, stats are outputted for every expression in the plan tree.
I have added a flag: `ignore-tables` that allows a set of stats tables
to be specified for which stats should not be outputted. Note that
care should be taken when rewritting stats for a test that uses the
`ignore-tables` flag, since the expressions that are being 'ignored'
may change if the plan changes.

Release note: None

Co-authored-by: Drew Kimball <drewk@cockroachlabs.com>

Commit:
Merge #52783

52783: bulkio: Fix racy test r=miretskiy a=miretskiy

Fixes #52707

Fix a race in the test, where the previously created
backup schedules would still continue executing, creating more backup
directories, which would cause cleanup function to fail.

Release Notes: None

Co-authored-by: Yevgeniy Miretskiy <yevgeniy@cockroachlabs.com>

Commit:
Merge #52588

52588: backupccl: enable backing up and restoring user defined schemas r=rohany a=rohany

Fixes #50875.

This commit adds support for backup/restore to operate when user defined
schemas have been created. We don't add a new `BACKUP SCHEMA` command,
as this would most likely create backups that have dangling references
across schemas.

Release note (enterprise change): Enable backup and restore when user
defined schemas are in use.

Co-authored-by: Rohan Yadav <rohany@alumni.cmu.edu>

Commit:
Merge #52739

52739: kvserver: add deadline to ClearRange r=nvanbenschoten a=dt

In many cases we would like to quickly and cheaply empty a key-range
and then make it available for SQL traffic. ClearRange is good at
quickly and cheaply emptying a key-range, but currently fears that
a copy of a ClearRange request could get replayed later make us fear
using it on a span which will later be sent writes -- if that replay
were to happen after writes resume, we could accidentally clear those
writes.

This change adds a Deadline parameter to ClearRange which the caller
can set to a timestamp to instruct the server to refuse to execute the
ClearRange request at or after that time. Thus if a caller sends the
ClearRanges with some deadline X, it can then wait until X (plus the
max clock offset) and then resume writing to the span it cleared. If
for any reason one of its ClearRange requests got stuck and a copy
pops out of a buffer after the deadline, it will have no effect.

Release note: none.

Co-authored-by: David Taylor <tinystatemachine@gmail.com>

spiffyy99 committed:
Merge branch 'tc-agent-filebeat' of github.com:r6min/cockroach into tc-agent-filebeat

r6min commented:
Closing to properly rebase into a new branch.
